Cleanup pyproject.toml file and dependencies

davidrudlstorfer opened the pull request:
No longer closes # 243 
No longer closes # 186

davidrudlstorfer commented:
I think the best way to tackle this is if we can discuss the changes. This would be a first proposal from my side regarding the dependency management.

Note that this is not a working version currently - I want to first clarify how we want to set it up - then fix the testing etc

What would be your take @isteinbrecher

isteinbrecher commented on pyproject.toml:
```diff
@@ -1,51 +1,58 @@
 [build-system]
-requires = ["setuptools", "wheel", "cython", "numpy"]
+requires = ["setuptools", "wheel"]
```
I am pretty sure we need python and numpy here, otherwise, we are not able to build the Cython code during the installation.

davidrudlstorfer replied:
I am currently investigating if we can completely decouple the cython stuff from a "regular" install - this way the cython code would only be compiled if we use `pip install . [cython]`

I think it would be good if MeshPy works as a pure Python library without any unnecessary overhead

isteinbrecher replied:
That would be great, this is also somewhat what I tried to say with my comment below. What I wanted to say here is that the package Cython is not really a dependency of MeshPy itself. We don't need it for MeshPy to run, we only need it for the code to be compiled during setup.

isteinbrecher commented on pyproject.toml:
```diff
@@ -1,51 +1,58 @@
-]
 description = "MeshPy: A beam finite element input generator"
+authors = [{ name = "MeshPy Authors" }]
+version = "0.1.0"
```
I am not that familiar with the version system here, but maybe we should use `dynamic = ["version"]`.

Also, we should discuss a versioning scheme at some point.

isteinbrecher commented on pyproject.toml:
```diff
@@ -1,51 +1,58 @@
+  "cython"
+]
+
+cubitpy = [
+  "cubitpy@git+https://github.com/imcs-compsim/cubitpy.git@main"
+]
```
I like having this as an optional dependency - if we do it like this we have to change the code so it won't fail if CubitPy is not available.

Also, if we add multiple optional packages, we should add a test case that tests the basic tests with just the "standard" installation.

isteinbrecher commented on pyproject.toml:
```diff
@@ -1,51 +1,58 @@
+  "pdoc"
 ]
 
+cython = [
+  "cython"
+]
```
The naming here is not ideal. At the moment the most performance critical part in MeshPy is the geometric search of matching points in space. We have three options there, a scipy implementation that works fine, the arborx implementation that is about twice as fast (but scales the same) as scipy and we have a brute force search written in cython that is the fastest for up until around 200 points.

So I could envision having some optional setup like (maybe with a better name):
`geometric_search_brute_force_cython`

And maybe also:
`geometric_search_bvh_arborx`
This would mean that we have to add the setup and installation of arborx to the meshpy package installation process (at the moment we do this manually during testing). Depending on the future of the ArborX implementation we can talk about that separately.

isteinbrecher reviewed (commented):
@davidrudlstorfer thanks for the initiative. I added some comments that we can talk about in the near future.

davidrudlstorfer commented:
One question - I now have a script which compiles the Cython Code independent from the install. I've also added a so-called script section to the pyproject.toml file. This way you can simply compile the Cython Code with `compile-cython`

Now one question - shouldn't a few tests fail if I do not compile the Cython Code? Somehow everything succeeds for me even if I did not compile the Cython Code @isteinbrecher

davidrudlstorfer commented:
> If the cython functionality is not available a few geometric search tests should fail.
> 
> Did you clean your repo? Cython creates the compiled binaries (I think right next to the cython code file) if this is still there from a previous installation it will be found.

Thanks for the information - I initially thought that it worked in another way and I only deleted the build folder - as you say the actual compiled file is directly next to the cython code file - thanks

davidrudlstorfer commented:
The current version now enables us to do the following:

- install MeshPy with different options, i.e., `pip install .[dev,cubitpy,python]` (note on zshell, i.e., macOS you have to escape the brackets like `pip install .\[dev,cubitpy,python\]`)

- install MeshPy either in noneditable mode or editable mode `pip install -e .` or `pip install .`

The entire setup now works - only things that are missing are the updates to the CICD and the actual source code to enable MeshPy without CubitPy and without the Cython code

davidrudlstorfer commented:
@isteinbrecher this would be ready for another round of review from you :partying_face: 

Overall all the things we wanted to achieve are now working as expected. We can now differentiate between editable and non editable installs. Also we can differentiate with all our dependencies (dev, cubitpy, cython).

The remaining ToDos would be:

- to finalize a testing suite we want to test. We should somewhere test the regular install + the editable install. And all different versions of dependencies.
- ideally we need to separately test the cython compilation and install
- finalize a name for the "cython" dependency
- update the readme to reflect all changes

Some points which will be helpful for review:

- Coverage only works in the editable mode since pytest-cov needs to access the source files. This is not possible in a non editable install (therefore, the cov is deactivated for the non editable installs)
- Unfortunately, I have no clue why 3 of the ArborX tests are failing? Maybe you can help me out here.
- We now need to run `compile-cython` after the install if we want to use it
- I excluded the compile-cython file from coverage because otherwise we would be below 90% for a normal test suite run

isteinbrecher reviewed (commented):
Thanks a lot for the work @davidrudlstorfer !

I had a look and left some comments. Regarding the optional dependencies: turns out the install of our Cython code works straight away with `setuptools` (see #258, I think the `src` layout helped a lot there). Since this works (and see the explanation below) I would stay with the approach "Install Cython code during build, allow for the user to optionally configure ArborX" and then reconsider this approach when actually creating the MeshPy PyPi package. Also, I would suggest to split the rest of this PR into a few smaller PR to go step-by-step. With this, the MeshPy module definition should be a lot improved, while still relying on `setuptools` as a build system.

1. Adapt `pyproject.toml` to allow for a working Cython code in editable and non-editable install with `setuptools`. I opened #258 which should hopefully show that this works almost straight away. I would also switch all test cases (except the one with 4C) to the installed variant.
2. Improve the handling of the "optional" CubitPy imports as proposed in this PR (including the changes to `environments.py`) @davidrudlstorfer could you do that?
3. Improve `pyproject.toml` (packages, metadata, ...) as proposed in this PR @davidrudlstorfer could you do that?
4. Improve the path finding for ArborX as proposed in this PR @davidrudlstorfer could you do that?

Did I miss anything?

Independent of that:
- Fix the ArborX implementation such that it allows for other than 3 and 6 dimensional points (I will take care of that)
- Maybe add the description of `geometric_search` below to the `__init__.py` file for the documentation?

___

The only drawback of the "Install Cython code during build" approach I can see at the moment is that this might lead to issues if no C compiler is available on the system (usually on Windows machines). If this turns out to be an issue during the creation of the PyPi package, we can drop the Cython support or have a look at the (https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#distributing-cython-modules or https://stackoverflow.com/questions/41778153/compiling-an-optional-cython-extension-only-when-possible-in-setup-py)

___

Maybe it makes sense to give some context to the `geometric_search` module. This module contains functionality to find points close to each other in a point cloud. This is the only real "submodule" of MeshPy, in the sense that it is an individual module that does not depend on the rest of MeshPy. Initially I wanted to create an own repo for this (still might do so) but at the moment I think it makes sense to keep it where it currently is. We have 3 different implementations for the actual search algorithm:
- brute force with cython: A brute force algorithm implemented in Cython, scales with $\mathcal{O}(n^2)$, but for a small number of points ($n<200$) this is the fastest algorithm since compared to the others it does not have any setup costs.
- bounding volume hierarchy (BVH) with scipy: Uses a BVH implementation provided by `scipy`. This scales with $\mathcal{O}(n\ \log{n})$.
- bounding volume hierarchy (BVH) with ArborX: Uses a BVH implementation provided by ArborX. This also scales with $\mathcal{O}(n\ \log{n})$ but due to a more optimised implementation is a few times faster than the `scipy` implementation.

The user should never worry about the actual algorithm, the `find_close_points` function should use the fastest (available!) algorithm for the given coordinate array. Because of that, and because of the modularisation argument above, I would not consider cython/arborx as "environments" of MeshPy (thus I would not add the cases to `environment.py`). The only case where we directly want to select the algorithm is during testing, simply to check that all implementations work.
If having 3 implementations turns out to be too much maintenance work, we could drop ArborX and Cython and just use the `scipy` implementation. This should have some small performance drawbacks, but in general should not affect the usability of MeshPy.

davidrudlstorfer commented on MANIFEST.in:
Good point - you are totally right. I've removed it

davidrudlstorfer commented on pyproject.toml:
```diff
@@ -1,57 +1,68 @@
 
+[tool.setuptools.packages.find]
+where = ["src"]
+include = ["meshpy"]
```
Also good point - it works as expected. And you are also right that pybind11 is not packaged because it has no __init__ file

davidrudlstorfer commented on pyproject.toml:
```diff
@@ -1,57 +1,68 @@
 [project]
 name = "MeshPy"
-authors = [
-  {name = "Ivo Steinbrecher", email = "ivo.steinbrecher@unibw.de"},
```
Good point - I've readded you as the main author and added the generic MeshPy Authors as a second line

davidrudlstorfer commented on src/meshpy/four_c/input_file.py:
```diff
@@ -45,6 +45,7 @@
 from meshpy.core.mesh import Mesh
 from meshpy.core.node import Node
 from meshpy.core.nurbs_patch import NURBSPatch
+from meshpy.utils.environment import is_cubitpy_available
```
Yes good idea - changed that

davidrudlstorfer commented:
First of all - thanks for your thorough review and also for questioning the setup as it has already improved it (removing the need to specify the package + the manifest file).

1. I've rebased to the latest main so I have the changes from your new ArborX implementation - it works as expected. Now the ArborX code is used for the cosserat curve and works 🥳 

2. I've adressed most of your points regarding your review and added them to this PR for now. I think it makes sense to split it up in a few chunks as you suggested

3. Where would you move the cython/arborx available functions? Into /meshpy/geometric_search/utils.py?

4. The only thing this review boils down to how we want to handle the Cython and CubitPy dependency (I think the rest is adressed and alright)

In my opinion we should really decouple everything as far as possible

> After thinking about it, it might be the best to just leave cubitpy in the dev dependencies, and maybe reconsider this when creating an actually MeshPy PyPI package. I am not sure if it makes sense to create a CubitPy package, then we can't use it as a direct dependency for MeshPy and might have to think about other ways to solve this anyway.

I think we should consider this already in this process because it plays a role regarding the import fixing etc. 
What do you mean with
> I am not sure if it makes sense to create a CubitPy package, then we can't use it as a direct dependency for MeshPy and might have to think about other ways to solve this anyway.

? Unfortunately, I do not understand this point.

The same goes for the Cython/Compile stuff. If we push the package to PyPI the user will download in 99% of the case via `pip install meshpy`. In this case the compiled cython code should be in the package - therefore, we do not need the cython install on the end users machine.

In my opinion it would probably make sense to move the cython dependency into the `dev dependencies` as it is only necessary to build the Cython Code as you've nicely summarized.

isteinbrecher commented:
> 3. Where would you move the cython/arborx available functions? Into /meshpy/geometric_search/utils.py?

I would actually leave that as it is. Each algorithm should already contain an error handling in case some compiled binaries can not be found. In this case I actually like the approach of checking everything when loading the package and storing the information on the availability in the variables `arborx_available` and `cython_available` (maybe they should be renamed to `GEOMETRIC_SEARCH_ARBORX_AVAILABLE` and `GEOMETRIC_SEARCH_CYTHON_AVAILABLE`). Is there anything I am missing?

> I am not sure if it makes sense to create a CubitPy package, then we can't use it as a direct dependency for MeshPy and might have to think about other ways to solve this anyway.

Sorry, I get that this was confusing 😅. My point is, that at the moment I am not planing on providing a `CubitPy` package on PyPi. Therefore, we can't use it as a direct "dependency" once MeshPy is published in PyPi. However, I would still like to maintain the compatibility of MeshPy with CubitPy. I think the best approach is to consider the CubitPy dependency as "We support it, but the user has to install it on his/her own" (similar to ArborX). To still test the compatibility we would simply install CubitPy "manually" in the testing (the CubitPy tests are already optional).

> The same goes for the Cython/Compile stuff. If we push the package to PyPI the user will download in 99% of the case via `pip install meshpy`. In this case the compiled cython code should be in the package - therefore, we do not need the cython install on the end users machine.

Again, I think my comments were confusing there. I fully agree with you, and I think this is already achieved with #258. We "outsource" the Cython dependency to the build process. `setuptools` creates an own virtual environment where Cython is installed and used to create the C code which is then compiled to a binary. Only the binary is copied to the install folder (editable or non-editable). Cython itself is not installed or required to actually run MeshPy (have a look at the installed packages in the job https://github.com/imcs-compsim/meshpy/actions/runs/13492675693/job/37693393402?pr=711#step:4:513, Cython is not installed).

My suggestion is to first merge #258, this should provide a stable build process for the Cython code without the need for much change in the repo. Then we can process with the other parts of this PR (dependency renaming to `dev`, cleanup of `pyproject.toml`, ...). My hope is, that with #258 packaging should automatically include the Cython binaries because it is entirely handled with `setuptools`. If that turns out to cause problems, I would not waste much time but only have the scipy geometric search. The cython dependency would be changed to a "We support it, but the user has to install it on his/her own with `pip compile-cython`" type of dependency (there we can then directly use the script provided in this PR), or even remove it.

isteinbrecher commented on pyproject.toml:
```diff
@@ -1,57 +1,67 @@
+  {name = "MeshPy Authors"}
 ]
-description = "MeshPy: A beam finite element input generator"
 readme = "README.md"
```
Should we add `requires-python = ">= 3.9"`?

isteinbrecher commented on pyproject.toml:
```diff
@@ -1,57 +1,67 @@
 
 [project]
 name = "MeshPy"
+description = "MeshPy: A beam finite element input generator"
```
"A general purpose 3D beam finite element input generator"

Maybe it makes sense to check if we can unify these variants in the repo (not part of this PR)

davidrudlstorfer commented:
This PR can now be closed.

All changes were introduced in #258 (editable install), #259 (fix ArborX), #263 (cubitpy import), #264 (geometric search documentation), #280 (update pyproject.toml file) and #282 (improve geometric_search imports)

@isteinbrecher thanks for our discussion regarding this topic - I am very happy with the now final outcome. Once again this improved MeshPy by a lot!

If we should run into a problem in the future with the automatic install/compilation of the cython code we can reuse the script used in this PR.
